x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/HttpCertificateCommand.java

```diff
@@ -0,0 +1,1149 @@
+                terminal.println("The key size must be a positive integer");
+                return null;
+            }
+        });
+    }
+
+    private char[] readPassword(Terminal terminal, String prompt, boolean confirm) {
+        while (true) {
+            final char[] password = terminal.readSecret(prompt + " [<ENTER> for none]");
+            if (password.length == 0) {
+                return password;
+            }
+            if (CertificateTool.isAscii(password)) {
+                if (confirm) {
+                    final char[] again = terminal.readSecret("Repeat password to confirm: ");
+                    if (Arrays.equals(password, again) == false) {
+                        terminal.println("Passwords do not match");
+                        continue;
+                    }
+                }
+                return password;
+            } else {
+                terminal.println(Terminal.Verbosity.SILENT, "Passwords must be plain ASCII");
+            }
+        }
+    }
+
+    private CertificateTool.CAInfo readKeystoreCA(Path ksPath, FileType fileType, Terminal terminal) throws UserException {
+        final String storeType = fileType == FileType.PKCS12 ? "PKCS12" : "jks";
+        terminal.println("Reading a " + storeType + " keystore requires a password.");
+        terminal.println("It is possible for the keystore's password to be blank,");
+        terminal.println("in which case you can simply press <ENTER> at the prompt");
+        final char[] password = terminal.readSecret("Password for " + ksPath.getFileName() + ":");
+        try {
+            final Map<Certificate, Key> keys = CertParsingUtils.readKeyPairsFromKeystore(ksPath, storeType, password, alias -> password);
+
+            if (keys.size() != 1) {
+                if (keys.isEmpty()) {
+                    terminal.println(Terminal.Verbosity.SILENT, "The keystore at " + ksPath + " does not contain any keys ");
+                } else {
+                    terminal.println(Terminal.Verbosity.SILENT, "The keystore at " + ksPath + " contains " + keys.size() + " keys,");
+                    terminal.println(Terminal.Verbosity.SILENT, "but this command requires a keystore with a single key");
+                }
+                terminal.println("Please try again with a keystore that contains exactly 1 private key entry");
+                throw new UserException(ExitCodes.DATA_ERROR, "The CA keystore " + ksPath + " contains " + keys.size() + " keys");
+            }
+            final Map.Entry<Certificate, Key> pair = keys.entrySet().iterator().next();
+            return new CertificateTool.CAInfo((X509Certificate) pair.getKey(), (PrivateKey) pair.getValue());
+        } catch (IOException | GeneralSecurityException e) {
+            throw new ElasticsearchException("Failed to read keystore " + ksPath, e);
+        }
+    }
+
+    private CertificateTool.CAInfo readPemCA(Path certPath, Path keyPath, Terminal terminal) throws UserException {
```

Comment:
It's tricky, because we need to go right back to the top (in case the first file was wrong).
Since `requestPath` already checks the file exists & is readable, I'd like to defer this for now.